versioned_hdf5/subchunk_map.py

```diff
@@ -336,3 +674,861 @@ def as_subchunk_map(
         chunk_subidxs = chunk_subidxs[:idx_len]
 
         yield Tuple(*chunk_slices), arr_subidxs, chunk_subidxs
+
```

Comment:
Refactor ends here. Everything below this line is new code that's not been integrated yet.
I must point out that the code cythonizes very poorly. This can be improved in a later iteration.